add desktop app

anbraten opened the pull request:
TODO
- [ ] add Github action to release binaries (`.exe, .dmg, .appimage, .flatpack`)
- [ ] think about auto update https://www.electron.build/auto-update
- [ ] add check to front to be able to detect if openend in electron to show some special setting (key bindings)
- [ ] create ipc api to interact with deployment
- [ ] add options page to front to set shortcut for mute / video on off
- [ ] add ipc to show notifications (chat message?)

moufmouf commented:
Whoa! So cool :heart: An Electron app!

Completely :+1: on the concept!

nolway commented:
Maybe in the game front we need to add something like below to know if the icp is available and use icp from the game for example to catch all notifications :
```ts
let ipcAvailable;

    try{
        var ipc = require('ipc');
        ipcAvailable = true;
    }
    catch(e){
        ipcAvailable = false;
    }
```

anbraten commented:
@nolway `desktop/src/app/types.ts` I have already added a type for the window interface. It will allow us to check if the frontend is executed from inside electron by checking `window.WorkAdventureDesktopApi.desktop`. To send a notification you could use `window.WorkAdventureDesktopApi.notify('my text')`. It uses a preload script with the electron ipc-bus behind it. Is this what would you meant?

anbraten commented:
@moufmouf The basic version of the desktop app should now be ready for a review. Th CI is building some executables of the app which you can download from the Github actions page. For testing you need to deploy a frontend with the changes of this PR to be able to use the shortcuts. Did you already had a look on #1822. Maybe it would be a good idea to start with squashing  to not get all of my ~70 testing commits separately into develop 🙈 

For the MacOS executable you have to add CodeSigning in the long term as that is required by Apple: https://www.electron.build/code-signing

moufmouf commented:
@anbraten This is really impressive. Thanks a lot again.

I'm having a hell of a week and I won't have time to review this today alas.
I'll do my best to free some time by the end of the week and make some feedbacks.

I'll keep you posted!

lukashass commented:
@anbraten Right now I have to click on the map to be able to move with _WASD_ after selecting a server.

After clicking on a server icon, the `appView` should be focused automatically IMO.

anbraten commented:
@lukashass Thanks for the hint, should be working now.

moufmouf commented:
Hi @anbraten ,

Reading your PR is taking me more time than I expected (mostly because I learn Electron at the same time :) )
I still need to review most of the code (was on holiday last week).

Anyway, here are my first thoughts / comments:

I started testing the global shortcuts feature on Ubuntu. I tried to register "Ctrl-Shift-M" for "Mute". For some reason, the application registered "M" as a global shortcut (without Ctrl-Shift).
The shortcut works great but... the "M" key stops working for all other applications.
It took me a good 5 minutes to understand this was not a problem with my keyboard but with WorkAdventure ;)
We should probably forbid a number of "obvious" keys and make it very clear these shortcut can be used even out of the application.

Second thought: you did an absolute huge job here. Thanks again! There is still a big amount of work to do to reach a 100% ready for production software. But I don't think we should wait for the PR to be 100% perfect to merge your work. It will make our life easier if we merge this PR soon enough, and then keep working on it until it is production ready (especially since auto-update is available!) We should probably mark the "WorkAdventure Desktop" application as "alpha" version to avoid being deceptive to our users (they can be quite demanding and provide us with a lot of feedback ;) )

Third thought: In the future, the application will need to connect to a variety of WorkAdventure servers, that could be in different versions. Because not everybody updates their WorkAdventure server to the latest version when a new version is released, some servers might not have all the IPC endpoints we could expect. As the communication between desktop app and WorkAdventure increases, we should keep this in mind, and maybe add a "getVersion" IPC endpoint to detect automatically the WorkAdventure version used on the currently open server.

anbraten commented:
Hi @moufmouf,

1. I had a similar situation where I used backspace as my shortcut and restarted my PC before realising it was captured by electron 🤦🏾. I implemented the Keyboard capturing for new shortcuts myself as I was searching two evenings for a library, but couldn't find any. The current approach (desktop/local-app/src/lib/KeyRecord.svelte) simply uses a normal input field taking all keyup events and waits for all inputs for 300ms. It somehow works, but is far from perfect and needs some tweaking in the long term.

2. Yeah that was my idea for this as well. I think it will be a nice base for us which we can evolve further. I like the idea to call it alpha for now and merge it soonishly as we then can also test the release cycle adjust it etc. I already added some other follow-up ideas as well to the description.

3. That's a nice idea as well. Currently you could enter the url of google and it as a server as well. I had the idea doing a get call to some unique url of the WA deployment and check if its is really a proper server as well. (Almost worked, but somehow I was unable to bundle the http with electron). Adding a version to the ipc interface would be just a few lines and should be no problem. We only need to think what should happen if the versions missmatch (showing a user info, simply disable some parts of the desktop-app, ...?).

Thanks for your thoughts and reviewing this PR.
